daemon/graphdriver/devmapper/deviceset.go

```diff
@@ -1744,10 +1752,99 @@ func (devices *DeviceSet) AddDevice(hash, baseHash string) error {
 		return fmt.Errorf("devmapper: device %s already exists. Deleted=%v", hash, info.Deleted)
 	}
 
-	if err := devices.createRegisterSnapDevice(hash, baseInfo); err != nil {
+	devinfo := &devInfo{}
+
+	if err := devices.parseStorageOpt(storageOpt, devinfo); err != nil {
+		return err
+	}
+
+	if err := devices.createRegisterSnapDevice(hash, baseInfo, devinfo.Size); err != nil {
 		return err
 	}
 
+	// Grow the container rootfs.
+	if devinfo.Size > 0 {
+		if err := devices.growFS(hash); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (devices *DeviceSet) parseStorageOpt(storageOpt []string, devinfo *devInfo) error {
+
+	// Read size to change the block device size per container.
+	var size int64
+	var err error
+	for _, option := range storageOpt {
+		val := strings.SplitN(option, ":", 2)
+		key := strings.ToLower(val[0])
+		switch key {
+		case "size":
+			size, err = units.RAMInBytes(val[1])
+			if err != nil {
+				return err
+			}
+			devinfo.Size = uint64(size)
+		default:
+			return fmt.Errorf("Unknown option %s\n", key)
+		}
+	}
+
+	return nil
+}
+
+func (devices *DeviceSet) growFS(hash string) error {
+
+	info, err := devices.lookupDevice(hash)
+	if info == nil {
+		return fmt.Errorf("Failed to lookup device %s:%v", hash, err)
+	}
+
+	info.lock.Lock()
+	defer info.lock.Unlock()
+
+	if err := devices.activateDeviceIfNeeded(info, false); err != nil {
+		return fmt.Errorf("Error activating devmapper device for '%s': %s", hash, err)
+	}
+
+	defer devices.deactivateDevice(info)
+
+	fsMountPoint := "/run/docker/mnt"
+	if _, err := os.Stat(fsMountPoint); os.IsNotExist(err) {
+		if err := os.MkdirAll(fsMountPoint, 0700); err != nil {
+			return err
+		}
+		defer os.RemoveAll(fsMountPoint)
+	}
+
+	options := ""
+	if devices.BaseDeviceFilesystem == "xfs" {
+		// XFS needs nouuid or it can't mount filesystems with the same fs
+		options = joinMountOptions(options, "nouuid")
+	}
+	options = joinMountOptions(options, devices.mountOptions)
+
+	if err := mount.Mount(info.DevName(), fsMountPoint, devices.BaseDeviceFilesystem, options); err != nil {
+		return fmt.Errorf("Error mounting '%s' on '%s': %s", info.DevName(), fsMountPoint, err)
+	}
+
+	defer syscall.Unmount(fsMountPoint, syscall.MNT_DETACH)
+
+	switch devices.BaseDeviceFilesystem {
+	case "ext4":
+		if out, err := exec.Command("resize2fs", info.DevName()).CombinedOutput(); err != nil {
+			logrus.Debugf("Failed to grow rootfs:%v:%s", err, string(out))
+		}
+	case "xfs":
+		if out, err := exec.Command("xfs_growfs", info.DevName()).CombinedOutput(); err != nil {
+			logrus.Debugf("Failed to grow rootfs:%v:%s", err, string(out))
+		}
+	default:
+		err = fmt.Errorf("Unsupported filesystem type %s", devices.BaseDeviceFilesystem)
```

Comment:
this error is never returned, you need to add `return err` at the end of the function.